test/dns-test.js

```diff
@@ -0,0 +1,396 @@
+      assert.equal(res.answer.length, 0);
+
+      // Handshake zone is not authoritative for ICANN based name
+      assert.equal(res.aa, false);
+    });
+
+    it('should return no answer for HNS name pre-update', async () => {
+      const name = 'foo.';
+
+      const res = await nstub.lookup(name, types.A);
+      assert.equal(res.answer.length, 0);
+    });
+
+    // record to save in the authenticated tree
+    for (const [hstype, item] of Object.entries(json)) {
+      const type = types[item.type];
+      const resource = Resource.fromJSON(item.resource);
+
+      // args: size, height, network
+      const name = rules.grindName(5, 1, network);
+
+      it(`should update ${name}`, async () => {
+        await wallet.client.post(`/wallet/${wallet.id}/open`, {
+          name: name
+        });
+
+        await mineBlocks(treeInterval + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/bid`, {
+          name: name,
+          bid: 1000,
+          lockup: 2000
+        });
+
+        await mineBlocks(biddingPeriod + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/reveal`, {
+          name: name
+        });
+
+        await mineBlocks(revealPeriod + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/update`, {
+          name: name,
+          data: resource.toJSON()
+        });
+
+        // mine a block
+        await mineBlocks(1, coinbase);
+
+        const rinfo = await nclient.execute('getnameresource', [name]);
+
+        const json = resource.getJSON(name);
+        assert.deepEqual(rinfo, json);
+
+        const info = await nclient.execute('getnameinfo', [name]);
+        const data = Buffer.from(info.info.data, 'hex');
+
+        const returned = Resource.decode(data);
+        assert.deepEqual(returned.getJSON(name), json);
+
+        // this commits the tree state to disk
+        // updates served by the dns server
+        await mineBlocks(treeInterval, coinbase);
+      });
+
+      it(`should return authenticated ${hstype} record`, async () => {
+        // Certain types of queries require the
+        // name to be formatted with additional data
+        const query = buildQuery(name, resource, type);
+
+        const response = await nstub.lookup(query, type);
+        const res = response.toJSON();
+        const dns = resource.toDNS(util.fqdn(name), type).toJSON();
+
+        // query for the zone signing key
+        const dnskey = await nstub.lookup('.', types.DNSKEY);
+
+        // parse the zone signing key and
+        // key signing key out of the response
+        const {zsk, ksk} = getSigningKeys(dnskey);
+
+        assert(zsk instanceof wire.Record);
+        assert(ksk instanceof wire.Record);
+
+        // validate the signature on the ZSK
+        verifyDNSSEC(dnskey, ksk, types.DNSKEY, '.');
+        // validate the signature over the rrsets
+        verifyDNSSEC(response, zsk, type, query);
```

Comment:
OK I see you are testing the keys here, maybe it should be a separate `it` block, or even run in `before()`

Reply:
It is testing that the zsk was signed by the ksk. I wouldn't trust that the zsk is correct until I know that it has been signed by the ksk, so that is why I test that it is signed each time. In actuality, it is not possible for the zsk to be man in the middle'ed in this test, so it could be done once. In practice, each request should test this, and as end to end tests I'd like to emulate as close as possible what would happen in reality. If it were more computationally complex I'd definitely agree.

If you ask me to refactor it so that it doesn't bid on a new name each time but instead uses a single name and just updates that name, I'd be happy to do that refactor. It slows down the tests quite a bit, but I didn't want any side effects from previous tests to have a role.